File: 6.Software-Design-Patterns/example/Structural/Proxy/Proxy.js
class Subject {
    request(){}
}
class RealSubject extends Subject {
    request(){
        console.log('RealSubject: Handling request.');
    }
}

class ProxyExample extends Subject {
    constructor(realSubject) {
        super()
        this.realSubject = realSubject;
    }

    request(){
        if (this.checkAccess()) {
            this.realSubject.request();
            this.logAccess();
        }
    }

    checkAccess(){
        // Some real checks should go here.
        console.log('Proxy: Checking access prior to firing a real request.');

        return true;
    }

    logAccess(){
        console.log('Proxy: Logging the time of request.');
    }
}

function clientCode(subject) {
    subject.request();
}
console.log('Client: Executing the client code with a real subject:');
const realSubject = new RealSubject();
clientCode(realSubject);
console.log('Client: Executing the same client code with a proxy:');
const proxy = new ProxyExample(realSubject);
clientCode(proxy);

File: 6.Software-Design-Patterns/example/Structural/Proxy/Proxy.go
package main

import (
	"fmt"
)

// Subject defines the interface for interacting with both RealSubject and Proxy
type Subject interface {
	Request()
}

// RealSubject represents the actual object with the requested functionality
type RealSubject struct{}

func (r *RealSubject) Request() {
	fmt.Println("RealSubject: Handling Request")
}

// Proxy acts as an intermediary for RealSubject
type Proxy struct {
	realSubject *RealSubject
}

// NewProxy creates a new Proxy instance with a reference to the RealSubject
func NewProxy(realSubject *RealSubject) *Proxy {
	return &Proxy{realSubject: realSubject}
}

// Request performs access control or additional tasks before delegating to RealSubject
func (p *Proxy) Request() {
	fmt.Println("Proxy: Checking access...")
	// Implement access control logic here (currently a placeholder)

	fmt.Println("Proxy: Forwarding request to RealSubject")
	p.realSubject.Request()
}

// clientCode takes a Subject interface and calls its Request method
func clientCode(subject Subject) {
	subject.Request()
}

func main() {
	fmt.Println("Client: Executing the client code with a real subject:")
	realSubject := &RealSubject{}
	clientCode(realSubject)

	fmt.Println("\nClient: Executing the same client code with a proxy:")
	proxy := NewProxy(realSubject)
	clientCode(proxy)
}
